x/cash/dynamicfee_test.go

```diff
@@ -0,0 +1,244 @@
+		t.Run(testName, func(t *testing.T) {
+			auth := helpers.Authenticate(tc.signers...)
+			bucket := NewBucket()
+			ctrl := NewController(bucket)
+			h := NewDynamicFeeDecorator(auth, ctrl)
+
+			tx := &txMock{info: &FeeInfo{Fees: &tc.txFee}}
+
+			db := store.MemStore()
+
+			gconf.SetValue(db, GconfCollectorAddress, collectorAddr)
+			gconf.SetValue(db, GconfMinimalFee, tc.minimumFee)
+
+			ensureWallets(t, db, tc.initWallets)
+
+			cache := db.CacheWrap()
+
+			cRes, err := h.Check(nil, cache, tx, tc.handler)
+			if !errors.Is(tc.wantCheckErr, err) {
+				t.Fatalf("got check error: %v", err)
+			}
+			if tc.wantGasPayment != cRes.GasPayment {
+				t.Errorf("gas payment: %d", cRes.GasPayment)
+			}
+
+			assertCharged(t, cache, ctrl, tc.wantCheckTxFee)
+
+			ensureWallets(t, cache, tc.updateWallets)
+
+			// If the check failed, deliver must not be called.
+			if tc.wantCheckErr != nil {
+				return
+			}
+
+			cache.Discard()
+
+			if _, err = h.Deliver(nil, cache, tx, tc.handler); !errors.Is(tc.wantDeliverErr, err) {
+				t.Fatalf("got deliver error: %v", err)
+			}
+
+			assertCharged(t, cache, ctrl, tc.wantDeliverTxFee)
+		})
+	}
+}
+
+var helpers x.TestHelpers
+
+// ensureWallets persist state of given wallet objects in the database. If
+// a wallet already exist it is overwritten.
+func ensureWallets(t *testing.T, db weave.KVStore, wallets []orm.Object) {
+	t.Helper()
+
+	bucket := NewBucket()
+	for i, w := range wallets {
+		if err := bucket.Save(db, w); err != nil {
+			t.Fatalf("cannot set %d wallet: %s", i, err)
+		}
+	}
+}
+
+// assertCharged check that given account was charged according to the fee
+// configuration.
+func assertCharged(t *testing.T, db weave.KVStore, ctrl Controller, want x.Coin) {
+	t.Helper()
+
+	minimumFee := gconf.Coin(db, GconfMinimalFee)
+	collectorAddr := gconf.Address(db, GconfCollectorAddress)
+
+	switch chargedFee, err := ctrl.Balance(db, collectorAddr); {
+	case err == nil:
+		wantTx := x.Coins{&want}
+		if !wantTx.Equals(chargedFee) {
+			t.Errorf("charged fee: %v", chargedFee)
+		}
+	case errors.Is(errors.ErrNotFound, err):
+		if minimumFee.IsZero() {
+			// Minimal fee is zero so the collector account is zero
+			// as well (not even created). All good.
+		} else {
+			if want.IsZero() {
+				// This is a weird case when a transaction was
+				// submitted but the signer does not have
+				// enough funds to pay the minimum (anty spam)
+				// fee.
+			} else {
+				t.Error("no fee charged")
+			}
+		}
+	default:
+		t.Errorf("cannot check collector account balance: %s", err)
+	}
+}
+
+type txMock struct {
+	weave.Tx
+	FeeTx
+	info *FeeInfo
+}
+
+func (m *txMock) GetFees() *FeeInfo {
+	return m.info
+}
+
+// Declare a unique error that can be matched in tests. This error is declared
+// only in tests so there is no way it can be returned by the implementation by
+// an accident.
+var ErrTestingError = errors.Register(123456789, "testing error")
+
+type handlerMock struct {
```

Comment:
Is there nothing similar in `x.Helpers`?

I tried to make all generic test helpers there, it would be good to use code from there if available, and if nothing fits the bill to move this into x.Helpers

Reply:
I see `func (TestHelpers) ErrorHandler(err error) weave.Handler` doing most of this, maybe add another one that succeeds on Check and fails on Deliver. (nothing should fail check and succeed deliver).

Reply:
Do you want me to add another method to the helpers structure? How about moving the mock there instead?

I always wondered why `TestHelpers` is a structure. Instead it could be a separate package that provides functions and structures. Similar to how [`net/http/httptest` is structured](https://golang.org/pkg/net/http/httptest/). `TestHelpers` does not carry any state and it seem to be acting only as a namespace workaround.

`ErrorHandler` provides some of the functionality. It always returns an error. The mock that I wrote returns whatever is set - it can for example [succeed with the `Check` method and fail on `Deliver` call](https://github.com/iov-one/weave/blob/master/x/cash/dynamicfee_test.go#L106). You can also set custom result if that is required.

What should I do? :smile_cat:

Reply:
> What should I do?

We should make a new issue of course....

I agree with moving to a new package. I did use this just for namespacing, and since the imports are

```
"github.com/iov-one/weave"
"github.com/iov-one/weave/crypto"
"github.com/tendermint/tendermint/libs/common"
```

We can pull it up to the top-level like `coin` I guess.

We have `ErrorHandler` and we have `WriteHandler`, which writes data and can return an empty result. I think adding a `SuccessHandler` that returns predefined Check/Deliver responses would be good.

And maybe one that returns Response on check and error on deliver... don't know what to name it.

If you like this idea, I will create a new issue for cleanup with the test helpers

Reply:
:link:  https://github.com/iov-one/weave/issues/372